### DISCUSSION TITLE
How to convert GLTF graphics elements in an iModel Connector

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/community#39] [Q&A]
Hello @pmconne 
I am hoping you can help again with a new question. I am developing the FutureOn - iTwin connector. It currently has code to convert GLTF graphics objects into iModel graphics elements, but I now need to update this to handle recent changes. I am wondering if it is possible to use the iTwin readGltf function (https://www.itwinjs.org/reference/core-frontend/tiles/readgltf/)

I can see that this is a frontend facility, is it possible to use this in a backend 'connector-framework' application to read a GLTF graphic object and convert it into an iModel graphic element?

**Comment by pmconne** [comment]
Hi Peter,

`readGltf` is only appropriate for creating graphics, not extracting geometry. We never really wanted to write our own glTF parser in iTwin.js, and the one we have is not fully spec-compliant. It pre-dates the standardizations of glTF 1.0 and 2.0. We plan to remove it eventually in favor of using a library specifically designed for reading glTF. I suggest you do the same. Here are a couple of suggestions:

- [loaders.gl](https://loaders.gl/) has modular support for a variety of formats, including glTF. We use it in iTwin.js for certain use cases, and it works great.
- [glTF-Transform](https://github.com/donmccurdy/glTF-Transform) looks quite promising. I have not tried it yet, but he maintainer, Don McCurdy, is an expert in this area.

**Reply by PeterJBassett** [reply]
Hello Paul @pmconne - Ok, thanks you very much for your answer.

**Reply by PeterJBassett** [reply]
Hello Paul @pmconne - as a follow on question. We currently use the threejs GltfLoader to read the gltf data and convert it into a threejs object.

Is there any example code anywhere for converting a threejs object into an iModel geometry object?

**Reply by pmconne** [reply]
No.

**Reply by PeterJBassett** [reply]
Hello @pmconne - in which case, is there any example code anywhere of using loaders.gl to convert objects into iModel geometry?

**Reply by pmconne** [reply]
However you decode the glTF, turning it into geometry will involve creating `Polyface`s. We do that [here](https://github.com/iTwin/itwinjs-core/blob/7c04b53424b8429cf76db8fd717cf03f01f30f22/core/frontend/src/tile/GltfReader.ts#L736). This code was not written to serve as an example.

**Reply by PeterJBassett** [reply]
Thanks very much for your help. It is very much appreciated. 
My lack of knowledge of iTwin means I am really struggling to get things to work.

**Reply by PeterJBassett** [reply]
Hello @pmconne a follow up question - I am try to convert a GLTF model that has vertex colors into an iModel object, but cannot get the vertex colors to work. This is the code I have tried, can you help identify why it is not working? The imodel object doesn't get any color, and shows as white.
```
        if (vertexColors) {
          for (let i = 0; i < vertexColors.length; i += 4) {
            colors.push( ColorDef.computeTbgrFromComponents(vertexColors[i] * 255, vertexColors[i + 1] * 255, vertexColors[i + 2] * 255, 0) );
          } 
        }
...
        for (let i = 0; i < indicies.length; i += 3) {
          polyfaceBuilder.addTriangleFacet(
            [points[indicies[i]], points[indicies[i + 1]], points[indicies[i + 2]]],
            undefined, 
            undefined,
            vertexColors && [colors[indicies[i]], colors[indicies[i + 1]], colors[indicies[i + 2]]]
          );
        }
```

**Reply by pmconne** [reply]
Does the mesh actually contain more than one color? `Polyface` supports per-vertex color but we haven't implemented support for it in the display code because we've literally never encountered a multi-colored mesh in the wild.

**Reply by PeterJBassett** [reply]
The latest asset library in FieldTwin uses vertex colors and uv textures; and then uses texture blending to achieve the surface color.

**Reply by pmconne** [reply]
> The latest asset library in FieldTwin uses vertex colors

Specifically, does it contain vertices that have different colors from other vertices?

You could do per-vertex color by creating a texture containing the unique colors and assigning UVs to each vertex. But it sounds like you've *also* already got a texture, which rules out that option.

A second option would be to extract all triangles of each unique color into a set of separate meshes, but that's hardly ideal. The ideal would be for us to support per-vertex colors directly.

How specifically is the vertex and texture color blended? Is it just a straight ratio for the entire mesh?

**Reply by PeterJBassett** [reply]
Hello @pmconne thanks again for your reply. 
The texture blending in FieldTwin is done using a multiply function.
Is it possible that iTwin can be upgraded to support vertex colors?

**Reply by pmconne** [reply]
Yes, but I don't have a timeline. @danieliborra @markschlosseratbentley

**Reply by danieliborra** [reply]
That is not in our roadmap. I'd recommend discussing this with Joe Granville, as our list of critical issues is long.
